File: CSharp/SpaceVIL/SpaceVIL/Items/TextItem.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using SpaceVIL.Common;
using SpaceVIL.Core;

namespace SpaceVIL
{
    abstract internal class TextItem : Primitive
    {
        private String _itemText = "";

        private Font _font = DefaultsService.GetDefaultFont();

        static int count = 0;

        internal TextItem()
        {
            SetItemName("TextItem_" + count);
            SetBackground(Color.Transparent);
            SetWidthPolicy(SizePolicy.Expand);
            SetHeightPolicy(SizePolicy.Expand);
            count++;
        }

        internal TextItem(String text, Font font) : this()
        {
            if (text == null)
            {
                text = "";
            }
            _itemText = text;
            if (font != null)
            {
                _font = font;
            }
        }

        internal TextItem(String text, Font font, String name) : this(text, font)
        {
            SetItemName(name);
        }

        internal String GetItemText()
        {
            return _itemText;
        }

        internal void SetItemText(String itemText)
        {
            if (itemText == null)
            {
                itemText = "";
            }
            if (!_itemText.Equals(itemText))
            {
                _itemText = itemText;
                UpdateData();
            }
        }

        internal Font GetFont()
        {
            return _font;
        }
        internal void SetFont(Font font)
        {
            if (font == null)
            {
                return;
            }
            if (!_font.Equals(font))
            {
                _font = font;
                UpdateData();
            }
        }
        internal void SetFontSize(int size)
        {
            if (_font.Size != size)
            {
                _font = FontService.ChangeFontSize(size, _font);
                UpdateData();
            }
        }
        internal void SetFontStyle(FontStyle style)
        {
            if (_font.Style != style)
            {
                _font = FontService.ChangeFontStyle(style, _font);
                UpdateData();
            }
        }
        internal void SetFontFamily(FontFamily font_family)
        {
            if (font_family == null)
            {
                return;
            }
            if (_font.FontFamily != font_family)
            {
                _font = FontService.ChangeFontFamily(font_family, _font);
                UpdateData();
            }
        }

        public abstract void UpdateData();

        private Color _foreground = Color.Black; //default
        public Color GetForeground()
        {
            return _foreground;
        }
        public void SetForeground(Color foreground)
        {
            if (!_foreground.Equals(foreground))
            {
                _foreground = foreground;
            }
        }
        public void SetForeground(int r, int g, int b)
        {
            SetForeground(GraphicsMathService.ColorTransform(r, g, b));
        }
        public void SetForeground(int r, int g, int b, int a)
        {
            SetForeground(GraphicsMathService.ColorTransform(r, g, b, a));
        }
        public void SetForeground(float r, float g, float b)
        {
            SetForeground(GraphicsMathService.ColorTransform(r, g, b));
        }
        public void SetForeground(float r, float g, float b, float a)
        {
            SetForeground(GraphicsMathService.ColorTransform(r, g, b, a));
        }

        private ItemAlignment _textAlignment;
        public ItemAlignment GetTextAlignment()
        {
            return _textAlignment;
        }
        public void SetTextAlignment(ItemAlignment value)
        {
            if (!_textAlignment.Equals(value))
            {
                _textAlignment = value;
            }
        }
        public void SetTextAlignment(params ItemAlignment[] alignment)
        {
            if (alignment == null)
                return;
            ItemAlignment common = alignment.ElementAt(0);
            if (alignment.Length > 1)
            {
                for (int i = 1; i < alignment.Length; i++)
                {
                    common |= alignment.ElementAt(i);
                }
            }
            SetTextAlignment(common);
        }
    }
}

File: Java/SpaceVIL/spacevil/src/main/java/com/spvessel/spacevil/TextItem.java
package com.spvessel.spacevil;

import com.spvessel.spacevil.Flags.ItemAlignment;
import com.spvessel.spacevil.Flags.SizePolicy;
import com.spvessel.spacevil.Common.DefaultsService;

import java.awt.*;
import java.util.*;
import java.util.List;

abstract class TextItem extends Primitive {
    private String _itemText = "";

    private Font _font = DefaultsService.getDefaultFont();

    private static int count = 0;

    TextItem() {
        setItemName("TextItem_" + count);
        setBackground(new Color(0, 0, 0, 0));
        setWidthPolicy(SizePolicy.EXPAND);
        setHeightPolicy(SizePolicy.EXPAND);
        count++;
    }

    TextItem(String text, Font font) {
        this();
        if (text == null) {
            text = "";
        }
        _itemText = text;
        if (font != null) {
            _font = font;
        }
    }

    TextItem(String text, Font font, String name) {
        this(text, font);
        setItemName(name);
    }

    String getItemText() {
        return _itemText;
    }

    void setItemText(String itemText) {
        if (itemText == null) {
            itemText = "";
        }
        if (!_itemText.equals(itemText)) {
            _itemText = itemText;
            updateData();
        }
    }

    Font getFont() {
        return _font;
    }

    void setFont(Font font) {
        if (font == null) {
            return;
        }
        if (!_font.equals(font)) {
            _font = font;
            updateData();
        }
    }

    void setFontSize(int size) {
        if (_font.getSize() != size) {
            _font = GraphicsMathService.changeFontSize(size, _font); //new Font(_font.getFamily(), _font.getStyle(), size);
            updateData();
        }
    }

    void setFontStyle(int style) {
        if (_font.getStyle() != style) {
            _font = GraphicsMathService.changeFontStyle(style, _font); //new Font(_font.getFamily(), style, _font.getSize());
            updateData();
        }
    }

    void setFontFamily(String font_family) {
        if (font_family == null) {
            return;
        }
        if (!_font.getFamily().equals(font_family)) {
            _font = GraphicsMathService.changeFontFamily(font_family, _font); //new Font(font_family, _font.getStyle(), _font.getSize());
            updateData();
        }
    }

    public abstract void updateData();

    private Color _foreground = Color.BLACK; // default

    public Color getForeground() {
        return _foreground;
    }

    public void setForeground(Color foreground) {
        if (foreground != null && !_foreground.equals(foreground)) {
            _foreground = foreground;
        }
    }

    public void setForeground(int r, int g, int b) {
        setForeground(GraphicsMathService.colorTransform(r, g, b));
    }

    public void setForeground(int r, int g, int b, int a) {
        setForeground(GraphicsMathService.colorTransform(r, g, b, a));
    }

    public void setForeground(float r, float g, float b) {
        setForeground(GraphicsMathService.colorTransform(r, g ,b));
    }

    public void setForeground(float r, float g, float b, float a) {
        setForeground(GraphicsMathService.colorTransform(r, g, b, a));
    }

    private List<ItemAlignment> _textAlignment = new LinkedList<>();

    public List<ItemAlignment> getTextAlignment() {
        return _textAlignment;
    }

    public void setTextAlignment(ItemAlignment... value) {
        setTextAlignment(Arrays.asList(value));
    }

    public void setTextAlignment(List<ItemAlignment> list) {
        if (list != null && !_textAlignment.equals(list)) {
            _textAlignment = list;
        }
    }
}